File: Solutions/LinkedListCycle.cpp
// Leetcode #141: https://leetcode.com/problems/linked-list-cycle/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// Solution 1 (Fast & Slow Pointer) | O(n) time, O(1) space
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr) return false; // Cycle can't exist if head is NULL
        
        // Initialize a fast and slower pointer to the head
        ListNode* fast = head;
        ListNode* slow = head;
        
        // While the fast node's next and next next exist
        while (fast->next && fast->next->next) {
            // Move fast pointer twice and slow pointer once
            fast = fast->next->next;
            slow = slow->next;
            
            // If the fast and slower pointer point to the same node, then the fast pointer "lapped" the slower one, so there must be a cycle
            if (fast == slow) return true;
        }
            
        return false; // If we exited the loop, there must not have been a cycle
    }
};

File: Solutions/LinkedListCycle.py
# Leetcode #141: https://leetcode.com/problems/linked-list-cycle/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Solution 1 (2 Pointers) | O(n) time, O(1) space
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Create a fast and slow pointer to track position in the list
        fast, slow = head, head

        # Move the fast pointer twice and slow pointer once. If they collide, there must be a cycle
        while (fast and fast.next):
            fast = fast.next.next
            slow = slow.next

            if fast == slow:
                return True

        # Otherwise, there can't be a cycle
        return False
